realm/realm-library/src/main/java/io/realm/internal/Context.java

```diff
@@ -17,135 +17,22 @@
 package io.realm.internal;
 
 import java.lang.ref.ReferenceQueue;
-import java.util.ArrayList;
-import java.util.List;
 
+// Currently we free native objects in two threads, the SharedGroup is freed in the caller thread, others are freed in
```

Comment:
Destruction of SharedGroup (and hence Group and Table) is currently not thread-safe with respect to destruction of other accessors, you have two ensure mutual exclusion. This is also illustrated by the use of locks in the testcases @beeender mentions above. Explicit call of SharedGroup::close() or Table::detach() is also not thread-safe with respect to destruction of other accessors.